### PR TITLE
[hail] Introduce `ArrayFold2` to permit more deforesting

### PR DESCRIPTION
**Pull request by tpoterba**
Stacked on #6980 

**Problem 1**: functions like `min`/`max` need to do a length check,
preventing us from implementing them in a single primitive fold.

**Problem 2**: functions like `mean` use  a struct as the accumulator,
leading to allocation (!) every element.

**Solution**: make it possible to have multiple primitive
accumulators. This is `ArrayFold2`.

The node is different from `ArrayFold` in that it:
 * has as sequence of accumulators, not just one
 * has a sequence of seq ops, one for each accumulator. Each of these
   sequence ops can see all the accumulators, and will see the updated
   value from sequence operations with a smaller index.
 * has a result op, which is a function from accumulators to result.

By changing `min`/`max` to use ArrayFold2 and inlining these functions,
we can get a reasonable speedup on `split_multi_hts`:

#6980 (this PR's parent):

```
2019-09-03 07:11:16,374: INFO:     burn in: 42.33s
2019-09-03 07:11:56,085: INFO:     run 1: 39.71s
2019-09-03 07:12:34,916: INFO:     run 2: 38.83s
2019-09-03 07:13:14,087: INFO:     run 3: 39.17s
```

PR:
```
2019-09-03 07:32:10,416: INFO:     burn in: 38.03s
2019-09-03 07:32:39,237: INFO:     run 1: 28.82s
2019-09-03 07:33:07,778: INFO:     run 2: 28.50s
2019-09-03 07:33:35,997: INFO:     run 3: 28.21s
```

### CODE REVIEW

**Comment by tpoterba** [comment]
Ported `ArrayFunctions.mean` to ArrayFold2.

Benchmark:
```python
@benchmark
def table_range_means():
    ht = hl.utils.range_table(10_000_000, 16)
    ht = ht.annotate(m = hl.mean(hl.range(0, ht.idx % 1111)))
    ht._force_count()
```

Master:
```
2019-09-03 09:39:05,777: INFO: [1/1] Running table_range_means...
2019-09-03 09:40:52,557: INFO:     burn in: 106.78s
2019-09-03 09:42:34,333: INFO:     run 1: 101.78s
2019-09-03 09:44:14,982: INFO:     run 2: 100.65s
2019-09-03 09:45:53,590: INFO:     run 3: 98.61s
```

PR:
```
2019-09-03 09:47:26,110: INFO: [1/1] Running table_range_means...
2019-09-03 09:47:29,465: INFO:     burn in: 3.35s
2019-09-03 09:47:32,615: INFO:     run 1: 3.15s
2019-09-03 09:47:35,703: INFO:     run 2: 3.09s
2019-09-03 09:47:38,840: INFO:     run 3: 3.14s
```

**Comment by tpoterba** [comment]
parent PR is in, ready for review.

Note that the next step is replacing ArrayFold with ArrayFold2.

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala [hunk 1]
```diff
@@ -136,27 +126,29 @@ object ArrayFunctions extends RegistryFunctions {
...
+        val first = genUID()
+        val acc = genUID()
+        ArrayFold2(a,
+          FastIndexedSeq((acc, NA(t)), (first, True())),
+          value,
+          FastIndexedSeq(
+            If(Ref(first, TBoolean()), Ref(value, t), invoke(op, t, Ref(acc, t), Ref(value, t))),
+            False()
+          ),
+          Ref(acc, t))
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: Keeping the boolean state seems unnecessary. What about only doing a fold when the array is non-empty, like:
```scala
val t = -coerce[TArray](a.typ).elementType
val accum = genUID()
val value = genUID()
val aUID = genUID()
val aRef = Ref(aUID, a.typ)
val zVal = ArrayRef(a, I32(0))
val body = invoke(op, t, Ref(value, t), Ref(accum, t))

Let(aUID, a,
  If(ApplyComparisonOp(EQ(TInt32()), ArrayLen(aRef), I32(0)),
    NA(t),
    ArrayFold(aRef, zVal, accum, value, body)))
```

File: hail/src/main/scala/is/hail/expr/ir/IR.scala [hunk 2]
```diff
@@ -235,6 +235,16 @@ final case class ArrayFlatMap(a: IR, name: String, body: IR) extends IR {
 }
 final case class ArrayFold(a: IR, zero: IR, accumName: String, valueName: String, body: IR) extends IR
 
+object ArrayFold2 {
+  def apply(a: ArrayFold): ArrayFold2 = {
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/IR.scala]: Would it be possible to make `ArrayFold.apply` call this and get rid of `ArrayFold` completely?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/IR.scala]: this is part 2, but it's a larger refactor and wanted to isolate the new behavior in a single PR.

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: (as discussed -- wrapping in a Let prevents deforestation)

**Pull request merged**